wordpress-mobile/WordPress-iOS
# PostService and AbstractPostListViewController Refactor

Originally, the `PostService` implementation did a couple things in which it managed paging of posts itself. In some sense, this just about attempted to manage its own state by paging requests based on whatever oldest post happened to be in CoreData, and also with purging any posts that weren't a part of a sync or paging request.

The above was an issue while implementing a new endpoint within `PostService` for searching posts. Searching posts created an issue because resulting posts would break the assumed paging of `PostService` implementation.

The primary refactor of this PR is within `PostService` and separating the use of sync and paging parameters instead on users of `PostService` via a new object parameter class `PostServiceOptions`. 

The secondary refactor is within `AbstractPostListViewController` and its supporting files, due to how intertwined the previous `PostService` sync, paging, and purge implementation were. `AbstractPostListViewController` now handles paging on its own and filters posts according to the sync results. Particularly, posts that are older than the `oldestPostDate` are filtered out of the `NSFetchRequest` to ensure posts arbitrarily loaded into CoreData are not displayed in a paged list (search results).

Additional refactoring includes method naming cleanup, use of generics and a bit of code-cleanup for clarity of what's going on.

**Testing:**
1. Open "Blog Posts", posts should load normally and in order from newest to oldest.
2. Scroll down the bottom of the list, older posts should load in order.
3. Continuing scrolling to load old posts until a desirable depth is reached.
4. Select either "Everyone" or "Only" in the segmented control above.
5. Posts should behave as expected as in steps 1, 2, 3.
6. Switch back to the previous segmented control option, the post list should be at the top and paging should continue where you left off for either segment control option.
7. Select a post to view it, go back, post list should maintain scroll position except for significant updates of ordering (unlikely).
8. Create a new post and post it, post should show up at the top of the list.
9. Select different post filters from the drop down, posts should display, page and order as expected.
10. Go back to the Site menu and select "Pages". Everything should work as expected 😎.

Please review @aerych 💥💥💥💥💥
